### PR TITLE
DRAFT: Adds support for C++ parallel algorithms

### PR DESCRIPTION
**Pull request by jefflarkin**
Support added for the C++ standard parallel algorithms. The changes have been tested with NVHPC 21.11 on both multicore CPU and GPU. To build with GPU, add `-DSTDPAR_FLAGS='-stdpar=gpu'`. To build for multicore CPU, add `-DSTDPAR_FLAGS='-stdpar=multicore'`. I have not tested with other Parallel C++ implementations. 

The NVIDIA implementation for GPUs relies on unified virtual memory memory. Several functions had to be modified to pass in data structures or make local copies of global scalars so that they would work both with multicore and GPU targets. There are likely other (better?) ways to handle this.

Because the C++ cartesian product is not yet available, I had to pack and unpack the loop indices myself using a couting_iterator. When ranges and cartesian_product are fully supported, they will be a nicer solution. 

Note: I based this on the version from the C directory rather than a version from the C++ directory to avoid a dependency on YAKL.

### CODE REVIEW

**Comment by mrnorman** [comment]
Change CMake lists to use semicolon rather than space.

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -113,9 +113,39 @@ if ("${ARCH}" STREQUAL "CUDA")
...
+endif()
+set(_COMPILER_FLAGS "${EXE_DEFS} -I${PNETCDF_PATH}/include ${STDPAR_FLAGS}")
+string(STRIP "${_COMPILER_FLAGS}" _COMPILER_FLAGS)
+set_target_properties(mpistdpar PROPERTIES COMPILE_FLAGS "${_COMPILER_FLAGS}")
+set(_LINK_FLAGS "-L${PNETCDF_PATH}/lib -lpnetcdf ${STDPAR_FLAGS}")
+string(STRIP "${_LINK_FLAGS}" _LINK_FLAGS)
+target_link_libraries(mpistdpar "${_LINK_FLAGS}")
+
+add_executable(mpistdpar_test miniWeather_mpi_stdpar.cpp)
+if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "PGI" AND "${STDPAR_FLAGS}" STREQUAL "")
```
Comment (jefflarkin) [inline cpp/CMakeLists.txt:140]: I don't like this anymore, maybe we take out this default or detect it better.

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -113,9 +113,39 @@ if ("${ARCH}" STREQUAL "CUDA")
...
+  set(STDPAR_FLAGS "-stdpar=gpu")
+endif()
+set(_COMPILER_FLAGS "${TEST_DEFS} -I${PNETCDF_PATH}/include ${STDPAR_FLAGS}")
+string(STRIP "${_COMPILER_FLAGS}" _COMPILER_FLAGS)
+set_target_properties(mpistdpar_test PROPERTIES COMPILE_FLAGS "${_COMPILER_FLAGS}")
 
+set(_LINK_FLAGS "-L${PNETCDF_PATH}/lib -lpnetcdf ${STDPAR_FLAGS}")
+string(STRIP "${_LINK_FLAGS}" _LINK_FLAGS)
+target_link_libraries(mpistdpar_test yakl "${_LINK_FLAGS}")
```
Comment (jefflarkin) [inline cpp/CMakeLists.txt:149]: Doesn't need yakl, this is a mistake.

File: cpp/miniWeather_mpi_stdpar.cpp [hunk 3]
```diff
@@ -0,0 +1,941 @@
...
+
+//Set this MPI task's halo values in the z-direction. This does not require MPI because there is no MPI
+//decomposition in the vertical direction
+void set_halo_values_z( double *state ) {
+  const double mnt_width = xlen/8;
+  int _nx = nx,
+      _nz = nz,
+      _i_beg = i_beg;
+  double _dx = dx;
+  thrust::counting_iterator<int> begin(0),end(NUM_VARS*(nx+2*hs));
```
Comment (mrnorman) [inline cpp/miniWeather_mpi_stdpar.cpp:469]: Might need to revisit this.

**Comment by mrnorman** [comment]
Might want to consider https://thrust.github.io/doc/classthrust_1_1zip__iterator.html

**Comment by jefflarkin** [comment]
I'm withdrawing this one in favor of an newer branch. I will open another PR in the near future.

**Pull request closed without merging**